File: string/longestSubstringWithoutRepeatingChars.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        map = [0] * 128
        start = end = 0
        result = 0

        while end < len(s):
            c1 = ord(s[end])
            map[c1] += 1

            while map[c1] > 1:
                c2 = ord(s[start])
                map[c2] -= 1
                start += 1

            result = max(result, end-start+1)
            end += 1

        return result


s = Solution()
answer = s.lengthOfLongestSubstring('abca bcbb')
print(answer)

File: string/longestSubstringWithoutRepeatingChars.js
const solution = {
  lengthOfLongestSubstring: (s) => {
    let dict = {};
    let longest = 0;
    let start = 0,
      end = 0;

    if (s.length === 1) {
      return 1;
    }

    while (end < s.length) {
      let letter = s.charAt(end);
      if (!dict[letter]) {
        dict[letter] = 1;
      } else {
        dict[letter] += 1;
      }
      while (dict[letter] > 1) {
        dict[s.charAt(start)] -= 1;
        start++;
      }
      longest = Math.max(longest, end - start + 1);
      end++;
    }
    return longest;
  },
};

const answer = solution.lengthOfLongestSubstring("pwwkew");
console.log(answer);
